### PR TITLE
196 Potential race condition in read-modify-write PDF vectors in parallel loop in onesource::generatePDF()

### PR DESCRIPTION
**Pull request by johannct**
## Change Description

- [ ] My PR includes a link to the issue that I am addressing

## Code Quality
- [ ] I have read the Contribution Guide
- [ ] My code follows the code style of this project
- [ ] My code builds (or compiles) cleanly without any errors or warnings
- [ ] My code contains relevant comments and necessary documentation

### Bug Fix Checklist
- [ ] My fix includes a new test that breaks as a result of the bug (if possible)
- [ ] My change includes a breaking change
  - [ ] My change includes backwards compatibility and deprecation warnings (if possible)

### New Feature Checklist
- [ ] I have added or updated the docstrings associated with my feature using the [NumPy docstring format](https://numpydoc.readthedocs.io/en/latest/format.html)
- [ ] I have updated the tutorial to highlight my new feature (if appropriate)
- [ ] I have added unit/End-to-End (E2E) test cases to cover my new feature
- [ ] My change includes a breaking change
  - [ ] My change includes backwards compatibility and deprecation warnings (if possible)

### Documentation Change Checklist
- [ ] Any updated docstrings use the [NumPy docstring format](https://numpydoc.readthedocs.io/en/latest/format.html)

### Build/CI Change Checklist
- [ ] If required or optional dependencies have changed (including version numbers), I have updated the README to reflect this
- [ ] If this is a new CI setup, I have added the associated badge to the README

### Other Change Checklist
- [ ] Any new or updated docstrings use the [NumPy docstring format](https://numpydoc.readthedocs.io/en/latest/format.html).
- [ ] I have updated the tutorial to highlight my new feature (if appropriate)
- [ ] I have added unit/End-to-End (E2E) test cases to cover any changes
- [ ] My change includes a breaking change
  - [ ] My change includes backwards compatibility and deprecation warnings (if possible)

### CODE REVIEW

File: src/lib/onesource.cpp [hunk 1]
```diff
@@ -1048,16 +1048,16 @@ void onesource::generatePDF(vector<SED *> &fulllib, const vector<size_t> &va,
...
-  pdfbayzq.vPDF.assign((PDFzqloc), (PDFzqloc+pdfbayzq.size()));
-  pdfmass.vPDF.assign((PDFmassloc), (PDFmassloc+pdfmass.size()));
-  pdfsfr.vPDF.assign((PDFSFRloc), (PDFSFRloc+pdfsfr.size()));
-  pdfssfr.vPDF.assign((PDFsSFRloc), (PDFsSFRloc+pdfssfr.size()));
-  pdfage.vPDF.assign((PDFAgeloc), (PDFAgeloc+pdfage.size()));
-  pdfldust.vPDF.assign((PDFLdustloc), (PDFLdustloc+pdfldust.size()));
-  pdfcol1.vPDF.assign((PDFcol1loc), (PDFcol1loc+pdfcol1.size()));
-  pdfcol2.vPDF.assign((PDFcol2loc), (PDFcol2loc+pdfcol2.size()));
-  pdfmref.vPDF.assign((PDFmrefloc), (PDFmrefloc+pdfmref.size()));
+  pdfbayzg.vPDF.assign((PDFzloc), (PDFzloc + pdfbayzg.size()));
```
Comment (hdante) [inline src/lib/onesource.cpp:1051]: This is a coding style change unrelated to the pull request, should it have a different commit ?

**Review by hdante — COMMENTED** [review]
Overall: this pull request was linked to issue #197 but it fixes #196 instead.

File: src/lib/onesource.cpp (hunk 1 above)
Comment (johannct) [reply src/lib/onesource.cpp:1051]: It is done automatically by pre-commit, and comes from a previous merge from you that I forced, so I think we can get it through here now.

**Comment by johannct** [comment]
> Overall: this pull request was linked to issue #197 but it fixes #196 instead.

damn..... I am not sure how to solve this....

File: CMakeLists.txt [hunk 2]
```diff
@@ -64,11 +64,23 @@ else()
...
+endif()
+
+
 # only set these flags if in `INCLUDE_COVERAGE` env var is set (to any value)
+# on linux/gcc it seems that one needs to run with LD_PRELOAD=/lib/x86_64-linux-gnu/libtsan.so.0 prepended
 if(DEFINED ENV{INCLUDE_COVERAGE})
   message(STATUS "Building with code coverage enabled.")
-  SET(GCC_COVERAGE_COMPILE_FLAGS "-g -O0 -coverage -fprofile-arcs -ftest-coverage")
-  SET(GCC_COVERAGE_LINK_FLAGS    "-coverage -lgcov")
+  SET(GCC_COVERAGE_COMPILE_FLAGS "-g -O0 -coverage -fprofile-arcs -ftest-coverage -fPIC -pie")
```
[the remark this answers is not in the document]
Comment (johannct) [reply CMakeLists.txt]: agreed. Fixed in #7aee2ff

File: CMakeLists.txt [hunk 3]
```diff
@@ -64,11 +64,23 @@ else()
...
 endif()
 
+# only set these flags if testing presence of race conditions :
+# https://github.com/google/sanitizers/wiki/ThreadSanitizerCppManual
+if(DEFINED ENV{SANITIZE_THREAD})
+  message(STATUS "Building with thread sanitizing enabled.")
+  SET(GCC_COVERAGE_COMPILE_FLAGS "-g -O2 -fsanitize=thread")
+  SET(GCC_COVERAGE_LINK_FLAGS    "-fsanitize=thread")
+  SET( CMAKE_CXX_FLAGS  "${CMAKE_CXX_FLAGS} ${GCC_COVERAGE_COMPILE_FLAGS}" )
+  SET( CMAKE_EXE_LINKER_FLAGS  "${CMAKE_EXE_LINKER_FLAGS} ${GCC_COVERAGE_LINK_FLAGS}" )
```
[the remark this answers is not in the document]
Comment (johannct) [reply CMakeLists.txt]: agreed, fixed in #7aee2ff

File: CMakeLists.txt [hunk 4]
```diff
@@ -64,11 +64,23 @@ else()
     add_compile_options("-mno-vzeroupper" "-mavx2" "-floop-unroll-and-jam" )
 endif()
 
+# only set these flags if testing presence of race conditions :
+# https://github.com/google/sanitizers/wiki/ThreadSanitizerCppManual
+if(DEFINED ENV{SANITIZE_THREAD})
+  message(STATUS "Building with thread sanitizing enabled.")
+  SET(GCC_COVERAGE_COMPILE_FLAGS "-g -O2 -fsanitize=thread")
+  SET(GCC_COVERAGE_LINK_FLAGS    "-fsanitize=thread")
+  SET( CMAKE_CXX_FLAGS  "${CMAKE_CXX_FLAGS} ${GCC_COVERAGE_COMPILE_FLAGS}" )
```
[the remark this answers is not in the document]
Comment (johannct) [reply CMakeLists.txt]: agreed, fixed in #7aee2ff

**Comment by hdante** [comment]
Hello, Johann, the conclusion for these patches is that they work now, but when enabling the thread sanitizer and testing, I'm receiving what at first look seem to be false positive data races caused by my OpenMP library, which seems to be Intel's library implicitly installed by conda. Are you having any data race warnings ? I think I'll just open another issue that might be closed as invalid if the false positive is confirmed. The main question is then if it makes sense to have an option to enable the thread sanitizer if it doesn't play well with OpenMP.

The patch now looks good to me, if this repository allows forced pushes, I suggest squashing the commits with fixups into the original patch (96ab9900238532fd810f68606998f7fa6517ca94) to avoid inserting commits with issues in the main branch.

https://stackoverflow.com/questions/33004809/can-i-use-thread-sanitizer-for-openmp-programs

**Comment by johannct** [comment]
Yes, I also saw a lot of warnings, starting with mag_gal, that I did not understand. A repeating one is the creation of oneSEDInt in each thread out of a given external single oneSED : the thread sanitizer seems to complain about copy constructor....
It does not cost much to leave the compilation option.
I did not understand what you have in mind with the last sentence about how to merge the fix.

**Comment by hdante** [comment]
Ok, so fully acknowledged. My last suggestion was to squash 3 of the proposed commits 96ab9900238532fd810f68606998f7fa6517ca94, 8f06ccf789be54243b95c5607720a28cae3276a9 and 7aee2ff852d7ce1e1ab9cc46fb3269e4896ecc35 into a single new commit containing the final proposed changes before merging them into the main branch.

**Review by hdante — APPROVED** [review]
Ack

**Comment by johannct** [comment]
@hdante dante looks like you need to redo a formal review :(

**Comment by johannct** [comment]
darn.... so my squashing did not proceed as planned? Yes let's do as you suggest.

**Comment by hdante** [comment]
Ok, created pull request https://github.com/lephare-photoz/lephare/pull/209

**Comment by johannct** [comment]
This has been dealt with more properly in another PR

**Pull request closed without merging**